File: src/RehabEasy.Infrastructure/Services/PatientHistoryReportBuilder.cs
using System.Text;
using RehabEasy.Domain.Models;

namespace RehabEasy.Infrastructure.Services;

public static class PatientHistoryReportBuilder
{
    public static string Build(PatientHistorySnapshot history)
    {
        StringBuilder builder = new();
        builder.AppendLine("RELATORIO DE HISTORICO DO PACIENTE - RehabEasy");
        builder.AppendLine($"Paciente ID: {history.PatientId}");

        if (!string.IsNullOrWhiteSpace(history.PatientName))
        {
            builder.AppendLine($"Nome: {history.PatientName}");
        }

        builder.AppendLine($"Gerado em: {DateTime.Now:g}");
        builder.AppendLine($"Total de testes: {history.Tests.Count}");
        builder.AppendLine($"Versoes de prontuario: {history.ClinicalNotes.Count}");
        builder.AppendLine();

        builder.AppendLine("=== HISTORICO DE TESTES ===");
        if (history.Tests.Count == 0)
        {
            builder.AppendLine("Nenhum teste registrado para este paciente.");
        }
        else
        {
            int index = 1;
            foreach (PatientTestHistoryEntry test in history.Tests.OrderByDescending(test => test.ReceivedAt))
            {
                builder.AppendLine();
                builder.AppendLine($"[{index}] {test.ReceivedAt.LocalDateTime:g} | {test.TestType} | {test.Title}");
                builder.AppendLine($"Indicadores: {test.MetricsSummary}");
                builder.AppendLine(test.DetailText);
                builder.AppendLine(new string('-', 72));
                index++;
            }
        }

        builder.AppendLine();
        builder.AppendLine("=== HISTORICO DE PRONTUARIO ===");
        if (history.ClinicalNotes.Count == 0)
        {
            builder.AppendLine("Nenhuma versao anterior de prontuario salva.");
        }
        else
        {
            int index = 1;
            foreach (PatientClinicalNoteHistoryEntry note in history.ClinicalNotes.OrderByDescending(note => note.SavedAt))
            {
                builder.AppendLine();
                builder.AppendLine($"[{index}] Salvo em {note.SavedAt.LocalDateTime:g}");
                builder.AppendLine(note.Content);
                builder.AppendLine(new string('-', 72));
                index++;
            }
        }

        builder.AppendLine();
        builder.AppendLine("=== PRONTUARIO ATUAL ===");
        if (history.CurrentClinicalNote is null || string.IsNullOrWhiteSpace(history.CurrentClinicalNote.Content))
        {
            builder.AppendLine("(Sem prontuario atual registrado.)");
        }
        else
        {
            builder.AppendLine(
                $"Ultima atualizacao: {history.CurrentClinicalNote.UpdatedAt.LocalDateTime:g}");
            builder.AppendLine(history.CurrentClinicalNote.Content);
        }

        return builder.ToString().TrimEnd();
    }
}

File: java/src/main/java/com/rehabeasy/service/PatientHistoryReportBuilder.java
package com.rehabeasy.service;

import com.rehabeasy.model.PatientClinicalNote;
import com.rehabeasy.model.PatientClinicalNoteHistoryEntry;
import com.rehabeasy.model.PatientHistorySnapshot;
import com.rehabeasy.model.PatientTestHistoryEntry;

import java.time.ZoneId;
import java.time.format.DateTimeFormatter;

public final class PatientHistoryReportBuilder {
    private static final DateTimeFormatter DATE_FORMAT =
            DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm").withZone(ZoneId.systemDefault());

    private PatientHistoryReportBuilder() {
    }

    public static String build(PatientHistorySnapshot history) {
        StringBuilder builder = new StringBuilder();
        builder.append("RELATORIO DE HISTORICO DO PACIENTE - RehabEasy\n");
        builder.append("Paciente ID: ").append(history.patientId()).append('\n');
        if (history.patientName() != null && !history.patientName().isBlank()) {
            builder.append("Nome: ").append(history.patientName()).append('\n');
        }
        builder.append("Gerado em: ").append(DATE_FORMAT.format(java.time.Instant.now())).append('\n');
        builder.append("Total de testes: ").append(history.tests().size()).append('\n');
        builder.append("Versoes de prontuario: ").append(history.clinicalNotes().size()).append("\n\n");

        builder.append("=== HISTORICO DE TESTES ===\n");
        if (history.tests().isEmpty()) {
            builder.append("Nenhum teste registrado para este paciente.\n");
        } else {
            int index = 1;
            for (PatientTestHistoryEntry test : history.tests()) {
                builder.append('\n');
                builder.append('[').append(index++).append("] ")
                        .append(DATE_FORMAT.format(test.receivedAt()))
                        .append(" | ").append(test.testType())
                        .append(" | ").append(test.title()).append('\n');
                builder.append("Indicadores: ").append(test.metricsSummary()).append('\n');
                builder.append(test.detailText()).append('\n');
                builder.append("-".repeat(72)).append('\n');
            }
        }

        builder.append("\n=== HISTORICO DE PRONTUARIO ===\n");
        if (history.clinicalNotes().isEmpty()) {
            builder.append("Nenhuma versao anterior de prontuario salva.\n");
        } else {
            int index = 1;
            for (PatientClinicalNoteHistoryEntry note : history.clinicalNotes()) {
                builder.append('\n');
                builder.append('[').append(index++).append("] Salvo em ")
                        .append(DATE_FORMAT.format(note.savedAt())).append('\n');
                builder.append(note.content()).append('\n');
                builder.append("-".repeat(72)).append('\n');
            }
        }

        builder.append("\n=== PRONTUARIO ATUAL ===\n");
        PatientClinicalNote current = history.currentClinicalNote();
        if (current == null || current.content().isBlank()) {
            builder.append("(Sem prontuario atual registrado.)");
        } else {
            builder.append("Ultima atualizacao: ")
                    .append(DATE_FORMAT.format(current.updatedAt())).append('\n');
            builder.append(current.content());
        }
        return builder.toString().stripTrailing();
    }
}
